lib/vdsm/storage/mpathconf.py

```diff
@@ -428,3 +432,95 @@ def read_metadata():
         return Metadata(REVISION_OLD, private)
 
     return Metadata(REVISION_MISSING, private)
+
+
+def check_local_config():
+    """
+    Check if 'user_friendly_names' is enabled in the multipathd configuration.
+    The configuration is obtained by running the multipathd command to show
+    the local configuration and capturing the output.
+    Parse the output and check that the contents are valid.
+
+    Raises:
+        exception.InvalidConfiguration: if any section have
+            'user_friendly_names' enabled. Contains the offending section name.
+    """
+    mpathconf = []
+    _parse_conf(multipath.show_config_local().split('\n'), mpathconf)
+    for conf in mpathconf:
+        failed_section, check_ok = _check_conf_attrs(conf.name, conf.attrs)
+        if not check_ok:
+            if failed_section is None:
+                failed_section = conf
+            raise exception.InvalidConfiguration(
+                reason=("'user_friendly_names' is enabled "
+                        f"in multipath configuration. Check '{CONF_FILE}'."),
+                section=failed_section)
+        # Keep looping to ensure 'user_friendly_names' is not enabled by
+        # any other device.
+
+
+def _check_conf_attrs(name, attributes):
+    """
+    Check that 'user_friendly_names' is disabled in the attributes section
+    of the parsed multipathd configuration.
+    Search is recursive.
+
+    Args:
+        name (str): name of the section to print.
+        attributes: attributes in the configuration grammar can contain
+            either nested attributes or key-value paris.
+
+    Returns:
+        Tuple[ConfSection, bool]:
+            ConfSection: used if any nested section have 'user_friendly_names'
+                enabled. Contains the offending section tuple.
+            bool: all 'user_friendly_names' are disabled
+    """
+    for attr in attributes:
+        if isinstance(attr, ConfPair):
+            if attr.key == 'user_friendly_names' and attr.value == 'yes':
+                return None, False
+        else:
+            assert isinstance(attr, ConfSection)
+            # Nested attributes (recursive call)
+            _, check_ok = _check_conf_attrs(
+                f"{name}.{attr.name}", attr.attrs)
+            if not check_ok:
+                return attr, check_ok
+    return None, True
+
+
+def _parse_conf(mpath_out, parsed_conf, depth=0):
+    """
+    Recursively parse multipath daemon configuration output.
+
+    Args:
+        mpath_out (List[str]): List of string lines containing the output
+            of the multipath configuration.
+        parsed_conf (List): List to append the resulting parsed configuration.
+        depth (int): Recursion depth.
+
+    Returns:
+        Number of lines parsed in this recursion.
+    """
+    i = 0
+    while i < len(mpath_out):
+        line = mpath_out[i].strip()
+        i += 1
```

Comment:
This is good way to do this in C, but we can make this much simpler using
iteration. Assuming the caller sent us an iterator over the lines of the config:

    for line in mpath_out:
        process line...

Reply:
This does not work if you want to jump iterations. That's why I used index-based iteration.
But with the iterator suggested above I can change this easily.

And for C, at least in C++ we have range-based loops available since C++11 :)